### ISSUE TITLE
DDP how to evaluate with custom metrics

### ISSUE THREAD

**Issue by thomas-happify** [huggingface/accelerate#57]
Hi there,

Is there a way to **evaluate the entire eval dataset** using customized metrics instead of `datasets.Metric` ?
I'm using similar code to [this](https://github.com/huggingface/transformers/blob/8d43c71a1ca3ad322cc45008eb66a5611f1e017e/examples/pytorch/summarization/run_summarization_no_trainer.py#L523-L552)
I'm fine tuning a T5 on multitask learning so I can't use the `metric` directly, because different `prefix` associate with different evaluation metrics.

Please advise
Thanks for your help!

**Comment by sgugger** [comment]
Hi there. You can accumulate the `decoded_preds` and `decoded_labels` in a big array (or the arrays you might need) then compute your metric outside of the loop on all your predictions and labels.

It's hard to be more specific without knowing the metric you want to compute.

**Comment by thomas-happify** [comment]
@sgugger  
Yes, I'm planning to accumulate `decoded_preds` and `decoded_labels`  and compute metrics outside the loop.
I simply used `all_preds.append(decoded_labels)` & `all_labels.append(decoded_labels)`, but it causes system hang after 7 iterations and the whole script freezes.

```
        all_preds = []
        all_labels = []
        for step, batch in enumerate(eval_dataloader):
            with torch.no_grad():
                # ...
                decoded_preds = tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
                decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
                
                all_preds.append(decoded_preds)
                all_labels.append(decoded_labels)
                # after 7 iterations, it freezes here and never exit the loop

        result = compute_metrics(all_preds, all_labels) # never reaches this line

```

Thanks a lot for your help!

**Comment by thomas-happify** [comment]
@sgugger  I figured it out.
Thanks for the help again!
